README.md
# emacs-incremental-search

This extension provides emacs-like incremental search function.

This extension is forked from [alt-core's fork](https://github.com/alt-core/vsc-incremental-search) [ of siegebell's "Incremental search for multiple cursors" extension](https://github.com/siegebell/vsc-incremental-search).

## Features

### Forward search
- Default binding: `ctrl+s`
- Command: `extension.incrementalSearch.forward`

Enters incremental search mode and subsequently advances each cursor to the next match, selecting it.

### Backwards search
- Default binding: `ctrl+r`
- Command: `extension.incrementalSearch.backward`

Enters incremental search mode and subsequently advances each cursor to the next match and selecting it, in the reverse direction.

### Expand selection by the next forward match
- Default binding: `ctrl+shift+s` (when in incremental search mode)
- Command: `extension.incrementalSearch.expand`

Instead of moving the selections to the next match, this adds the next matches to the existing selection.

### Expand selection by the next backwards match
- Default binding: `ctrl+shift+r` (when in incremental search mode)
- Command: `extension.incrementalSearch.backwardExpand`

Expands the selections with a backwards search.

### Toggle regular expressions vs plain-text search terms
- Default binding: `alt+r` (when in incremental search mode)
- Command: `extension.incrementalSearch.toggleRegExp`

The search term is interpreted as a regular expression when enabled, or as plain text otherwise.

### Toggle case sensitivity
- Default binding: `alt+c` (when in incremental search mode)
- Command: `extension.incrementalSearch.toggleCaseSensitivity`

Toggles whether the match should be case sensitive.

## Known Issues

## Release Notes

See CHANGELOG.md


CHANGELOG.md
# Change Log

### 0.0.9

clear selection when finishing incremental search

### 0.0.8

add centerText method defaulting to Ctrl+L

### 0.0.7

default to case sensitive when there are uppercase chars in the search

### 0.0.6

update README to reflect current state, remove some dead code

### 0.0.5

keep previous search term & reuse when starting search
add configuration for centering text in different ways

### 0.0.4

center on cursor when cancelling a selection

### 0.0.3

make ctrl+g cancel out of input box

### 0.0.2

fix broken case insensitive non-regexp search

### 0.0.1

Initial release of emacs-incremental-search.


src/SearchExpr.ts
import * as vscode from 'vscode';
let dbg = vscode.window.createOutputChannel("IncrementalSearch-SearchExpr");

function regexpReverseExec(re: RegExp, text: string, start?: number) {
  if (start === undefined)
    start = re.lastIndex;
  else if (start < 0)
    start = 0;
  // Make sure we do a global search
  re = re.global
    ? re
    : new RegExp(re.source, 'g' + re.flags);

  re.lastIndex = 0;
  let lastMatch: RegExpExecArray = null;
  let match: RegExpExecArray;
  let idx = re.lastIndex;
  while ((match = re.exec(text)) != null && match.index < start && match[0] !== '') {
    // if(match.length > 1 && match.index+match[1].length >= start)
    //   break;
    lastMatch = match;
    re.lastIndex = match.index + 1; // do not jump further than one character from the match
  }
  if (match != null && match.length == 0 && match[0] == '')
    return null;
  else
    return lastMatch;
}

function hasUpperCase(str) {
  return (/[A-Z]/.test(str));
}

export class SearchExpr {
  private re: RegExp;
  private lastIdx: number;
  constructor(private searchTerm: string, private multiline: boolean, private useRegExp: boolean, private caseSensitive: boolean) {
    if (useRegExp)
      this.re = new RegExp(searchTerm, 'g' + (caseSensitive ? '' : 'i') + (multiline ? 'm' : ''))
    this.lastIndex = 0;
  }

  public set lastIndex(value: number) {
    if (this.useRegExp)
      this.re.lastIndex = value;
    else
      this.lastIdx = value;
  }

  public get lastIndex(): number {
    if (this.useRegExp)
      return this.re.lastIndex;
    else
      return this.lastIdx;
  }

  public exec(text: string, reverse = false): { index: number } & Array<string> {
    if (this.useRegExp) {
      if (reverse)
        return regexpReverseExec(this.re, text);
      else
        return this.re.exec(text);
    } else {
      let idx;
      let srchTerm = this.searchTerm;
      // if we have toggled on case sensitivity, or there are any uppercase
      // characters in the search, then we want to be case sensitive.
      if (!(this.caseSensitive || hasUpperCase(srchTerm))) {
        srchTerm = srchTerm.toLowerCase();
        text = text.toLowerCase();
      }

      if (reverse)
        idx = text.substring(0, this.lastIndex + this.searchTerm.length - 1).lastIndexOf(this.searchTerm, this.lastIdx);
      else
        idx = text.indexOf(this.searchTerm, this.lastIdx);

      if (idx < 0)
        return null;

      return Object.assign([this.searchTerm], { index: idx });
    }
  }
}


src/InlineInput.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';

let dbg = vscode.window.createOutputChannel("IncrementalSearchInlineInput");
const INLINE_INPUT_CONTEXT = 'incrementalSearchInlineInput';

const subscriptions : vscode.Disposable[] = [];

export function disposeInlineInput() {
  subscriptions.forEach((d) => d.dispose());
}

function registerTextEditorCommand(commandId:string, run:(editor:vscode.TextEditor,edit:vscode.TextEditorEdit,...args:any[])=>void): void {
  subscriptions.push(vscode.commands.registerTextEditorCommand(commandId, run));
}

function registerCommand(commandId:string, run:(...args:any[])=>void): void {
  subscriptions.push(vscode.commands.registerCommand(commandId, run));
}

class Input {
  text: string;
  validateInput: (text: string) => string;
  resolve: (text: string) => void;
  reject: (reason: any) => void;
  constructor(options: {
    text: string,
    validateInput: (text: string) => string,
    resolve: (text: string) => void,
    reject: (reason: any) => void,  
  }) {
    this.text = options.text;
    this.validateInput = options.validateInput;
    this.resolve = options.resolve;
    this.reject = options.reject;
  }
}
var registeredTypeCommand = false;

registerTextEditorCommand('extension.incrementalSearch.backspace', onBackspace);
registerTextEditorCommand('extension.incrementalSearch.stopInline', onEscape);
registerTextEditorCommand('extension.incrementalSearch.complete', onComplete);

const editorInputs = new Map<vscode.TextEditor,Input>();

vscode.window.onDidChangeActiveTextEditor((editor: vscode.TextEditor) => {
  const hasEditor = editorInputs.has(editor);
  vscode.commands.executeCommand('setContext', INLINE_INPUT_CONTEXT, hasEditor);  
})



export async function showInlineInput(editor: vscode.TextEditor, value: string, validateInput: (text: string) => string) : Promise<string> {
  if(editorInputs.has(editor))
    throw "TextEditor already has an inline input."

  await vscode.commands.executeCommand('setContext', INLINE_INPUT_CONTEXT, true);

  const result = new Promise<string>((resolve, reject) => {
    const input = new Input({text: value, validateInput: validateInput, resolve: resolve, reject: reject});
    editorInputs.set(editor,input);
  });

  if(!registeredTypeCommand) {
    registeredTypeCommand = true;
    registerCommand('type', onType);
  }

  return result;
}

function onEscape(editor: vscode.TextEditor) {
  cancel(editor, 'user initiated escape')
}

function onComplete(editor: vscode.TextEditor) {
  complete(editor);
}

function onBackspace(editor: vscode.TextEditor) {
  const input = editorInputs.get(editor);
  if(input) {
    input.text = input.text.substr(0,input.text.length-1);
    input.validateInput(input.text);
  }
}

function onType(event: {text:string}) {
  const input = editorInputs.get(vscode.window.activeTextEditor);
  if(input) {
    input.text+= event.text;
    input.validateInput(input.text);
  }
  else
    vscode.commands.executeCommand('default:type', event);
}

export function cancel(editor: vscode.TextEditor, reason = '') {
  const input = editorInputs.get(editor);
  if(input) {
    editorInputs.delete(editor);
    input.resolve(undefined);
  }
  if(editorInputs.size == 0)
    vscode.commands.executeCommand('setContext', INLINE_INPUT_CONTEXT, false);  
}

export function complete(editor: vscode.TextEditor) {
  const input = editorInputs.get(editor);
  if(input) {
    editorInputs.delete(editor);
    input.resolve(input.text);
  }
  if(editorInputs.size == 0)
    vscode.commands.executeCommand('setContext', INLINE_INPUT_CONTEXT, false);  
}



src/Configuration.ts
import * as vscode from 'vscode';

type DecorateMatchCondition = 'never' | 'always' | 'multigroups';
type CenterText = 'always' | 'onCancel' | 'never';

interface Configuration {
  matchStyle: vscode.DecorationRenderOptions,
  centerText: CenterText,
  // when to show the style
  styleMatches: DecorateMatchCondition,
  selectionStyle: vscode.DecorationRenderOptions,
  matchDecoration: vscode.TextEditorDecorationType | null;
  selectionDecoration: vscode.TextEditorDecorationType | null;
}


let configuration: Configuration = {
  matchStyle: {
    dark: {
      border: '1pt white dashed',
    },
    light: {
      border: '1pt black solid',
    },
  },
  centerText: 'always',
  styleMatches: 'always',
  selectionStyle: {
    backgroundColor: 'rgba(0,0,255,0.5)',
    borderRadius: '50%',
    border: '1pt rgba(0,0,100,0.8) solid',
  },
  matchDecoration: null,
  selectionDecoration: null
};

export function activate() {
  loadConfiguration();
  vscode.workspace.onDidChangeConfiguration(loadConfiguration);
}

export function get() {
  return configuration;
}

function loadConfiguration() {
  configuration = Object.assign(configuration, vscode.workspace.getConfiguration("incrementalSearch"));

  if (configuration.selectionDecoration)
    configuration.selectionDecoration.dispose();
  configuration.selectionDecoration = null;
  configuration.selectionDecoration = vscode.window.createTextEditorDecorationType(configuration.selectionStyle);

  if (configuration.matchDecoration != null)
    configuration.matchDecoration.dispose();
  configuration.matchDecoration = null;
  configuration.matchDecoration = vscode.window.createTextEditorDecorationType(configuration.matchStyle);
}


src/SearchStatusBar.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';

export class SearchStatusBar implements vscode.Disposable {
  private title: vscode.StatusBarItem;
  private matchCase: vscode.StatusBarItem;
  private useRegExp: vscode.StatusBarItem;

  constructor(toggleCaseSensitive: string, toggleRegularExpressions: string) {
    this.title = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 10);
    this.matchCase = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 9);
    this.useRegExp = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 8);

    this.title.text = 'incremental search: ';
    this.matchCase.text = 'Aa';
    this.useRegExp.text = '.*';
    this.matchCase.command = toggleCaseSensitive;
    this.useRegExp.command = toggleRegularExpressions;
    this.matchCase.tooltip = 'Toggle case sensitivity.'
    this.useRegExp.tooltip = 'Toggle the use of regular expressions for search.'
  }

  public dispose() {
    this.title.dispose();
    this.matchCase.dispose();
    this.useRegExp.dispose();
  }

  public show() {
    this.title.text = 'incremental search: ';
    this.title.show();
    this.matchCase.show();
    this.useRegExp.show();
  }

  public hide() {
    this.title.hide();
    this.matchCase.hide();
    this.useRegExp.hide();
  }

  public indicateSyntaxError() {
    this.title.color = 'red';
  }

  public indicateNoMatch() {
    this.title.color = 'yellow';
  }

  public update(searchTerm: string, caseSensitive: boolean, useRegExp: boolean, options = { backward: false }) {
    this.title.text = 'incremental search: $(chevron-left)' + searchTerm + '$(chevron-right)';
    this.title.color = 'white';
    this.matchCase.color = caseSensitive ? 'white' : 'red';
    this.useRegExp.color = useRegExp ? 'white' : 'red';
  }
}


src/IncrementalSearch.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import { SearchExpr } from './SearchExpr';
import * as configuration from './Configuration';

const INCREMENTAL_SEARCH_CONTEXT = 'incrementalSearch';


export enum SearchDirection {
  forward,
  backward,
}

export interface SearchOptions {
  direction?: SearchDirection,
  caseSensitive?: boolean,
  useRegExp?: boolean,
  expand?: boolean,
  searchTerm?: string,
}

export const INITIAL_OPTIONS: SearchOptions = {
  searchTerm: '',
  direction: SearchDirection.forward,
  caseSensitive: false,
  useRegExp: true,
  expand: false,
}

export const DEFAULT_OPTIONS: SearchOptions = {
  searchTerm: undefined,              // use prior value
  direction: SearchDirection.forward, //
  caseSensitive: undefined,           // use prior value
  useRegExp: undefined,               // use prior value
  expand: false,                      //
}

function getOrDefault<T>(x: T, d: T) {
  return x === undefined ? d : x;
}


export class IncrementalSearch {
  private currentSelections: vscode.Selection[];
  private initialSelections: vscode.Selection[];
  private matchedRanges: vscode.Range[] = [];
  // when we do an incremental search and want to keep the previous selections,
  // no are no longer active and are placed in this list
  private aggregatedSelections: vscode.Selection[] = [];

  constructor(
    private editor: vscode.TextEditor,
    private options: SearchOptions = INITIAL_OPTIONS
  ) {
    this.applyOptions(options, INITIAL_OPTIONS);
    this.initialSelections = editor.selections;
    this.currentSelections = editor.selections;
  }

  /** Applies new options to the current options. If a propert is undefined, then we keep the prior value */
  public applyOptions(options: SearchOptions, defaults = this.options) {
    for (const key in INITIAL_OPTIONS)
      this.options[key] = getOrDefault(options[key], defaults[key]);
  }

  public cancelSelections() {
    this.editor.selections = this.initialSelections;
    if (configuration.get().centerText != "never") {
      IncrementalSearch.centerText();
    }
  }

  public static async centerText() {
    await vscode.commands.executeCommand("revealLine", {
      lineNumber: vscode.window.activeTextEditor.selection.start.line,
      at: "center"
    });
  }

  public get searchTerm() { return this.options.searchTerm; }
  public get useRegExp() { return this.options.useRegExp; }
  public get caseSensitive() { return this.options.caseSensitive; }
  public get direction() { return this.options.direction; }
  public get expandSelections() { return this.options.expand; }

  public advance(options: SearchOptions = DEFAULT_OPTIONS) {
    this.applyOptions(options);
    let nextPositions: vscode.Position[] = [];
    //this.initialSelections = this.matchedRanges.map((r) => new vscode.Selection(r.start,r.end));
    if (this.options.direction == SearchDirection.backward)
      for (const sel of this.matchedRanges) {
        nextPositions.push(sel.start);
      }
    else // forward search
      for (const sel of this.matchedRanges) {
        nextPositions.push(sel.start.translate(0, 1));
      }
    // if(nextSelections.length > 0)
    //   this.initialSelections = nextSelections;
    if (nextPositions.length > 0)
      return this.update(options, nextPositions);
    else
      return this.update(options);
  }

  public getSelections(): vscode.Selection[] {
    return normalizeSelections(this.aggregatedSelections.concat(this.currentSelections));
  }

  public getMatchedRanges() {
    return this.matchedRanges;
  }

  public getEditor() {
    return this.editor;
  }

  public update(options: SearchOptions = DEFAULT_OPTIONS, startingPositions?: vscode.Position[]): { matchedRanges: vscode.Range[], matchedGroups: boolean } {
    try {
      if (!startingPositions)
        startingPositions = this.initialSelections.map((sel) => sel.active);

      this.applyOptions(options);

      const text = this.editor.document.getText();
      const search = new SearchExpr(this.searchTerm, true, this.useRegExp, this.caseSensitive);
      var matchingGroups = false;
      let nextSelections = [];
      this.matchedRanges = [];

      if (options.expand)
        this.aggregatedSelections = normalizeSelections(this.aggregatedSelections.concat(this.currentSelections));
      else
        this.aggregatedSelections = [];

      for (const pos of startingPositions) {
        let start = this.editor.document.offsetAt(pos);
        const retry_position = this.direction == SearchDirection.forward ? 0 : this.editor.document.offsetAt(new vscode.Position(this.editor.document.lineCount, 0));
        while (true) {
          search.lastIndex = start;
          const match = search.exec(text, this.direction == SearchDirection.backward);
          if (match !== null && match.length > 0) {
            const newAnchor = this.editor.document.positionAt(match.index);
            const newActive = this.editor.document.positionAt(match.index + match[0].length);
            this.matchedRanges.push(new vscode.Range(newAnchor, newActive));
            if (match.length == 1) {
              nextSelections.push(new vscode.Selection(newAnchor, newActive));
            }
            else {
              // The regexp contains subgroups
              matchingGroups = true;
              // Turn each subgroup into a new selection
              let offset = 0;
              match
                .forEach((m: any, idx: any) => {
                  if (idx == 0 || m === undefined)
                    return; // skip the first element
                  offset = match[0].indexOf(m, offset);
                  const newAnchor = this.editor.document.positionAt(match.index + offset);
                  offset += m.length;
                  const newActive = this.editor.document.positionAt(match.index + offset);
                  nextSelections.push(new vscode.Selection(newAnchor, newActive));
                });
            }
            break;
          }
          else {
            if (start == retry_position) {
              break;
            }
            else {
              start = retry_position;
            }
          }
        }
      }
      if (nextSelections.length > 0)
        this.setEditorSelections(nextSelections);
      else {
        this.matchedRanges = this.initialSelections;
        this.setEditorSelections(this.initialSelections);
      }
      this.matchedRanges = normalizeRanges(this.matchedRanges);
      return { matchedRanges: this.matchedRanges, matchedGroups: matchingGroups }
    } catch (e) {
      this.setEditorSelections(this.initialSelections);
      throw e;
    }
  }

  private setEditorSelections(selections: vscode.Selection[]) {
    this.currentSelections = normalizeSelections(selections);
    if (this.aggregatedSelections.length + selections.length == 0)
      return;
    const normSelections = normalizeSelections(this.aggregatedSelections.concat(selections));
    this.editor.revealRange(selections[0]);
    this.editor.selections = normSelections;
    if (configuration.get().centerText == "always") {
      IncrementalSearch.centerText();
    }
  }

}

const normalizeRanges = (selections: vscode.Range[]) => normalizeRangesGeneric(selections, vscode.Range);

const normalizeSelections = (selections: vscode.Selection[]) => normalizeRangesGeneric(selections, vscode.Selection);
// const sorted = selections
// 	.slice(0,selections.length)
//   .sort((x,y) => x.start.isBefore(y.start) ? -1 : x.start.isAfter(y.start) ? 1 : x.end.isBefore(y.start) ? -1 : x.end.isAfter(y.end) ? 1 : 0);
// const results = [sorted.shift()];
// let currentIdx = 0;
// for(let idx = 0; idx < sorted.length; ++idx) {
// 	if(sorted[idx].start.isBeforeOrEqual(results[currentIdx].end))
// 		results[currentIdx] = new vscode.Selection(results[currentIdx].start,sorted[idx].end)
// 	else {
// 		results.push(sorted[idx]);
// 		++currentIdx;
// 	}
// }
// return results;

function normalizeRangesGeneric<T extends vscode.Range>(selections: T[], TT: { new(anchor: vscode.Position, active: vscode.Position): T }): T[] {
  const sorted = selections
    .slice(0, selections.length)
    .sort((x, y) => x.start.isBefore(y.start) ? -1 : x.start.isAfter(y.start) ? 1 : x.end.isBefore(y.start) ? -1 : x.end.isAfter(y.end) ? 1 : 0);
  const results = [sorted.shift()];
  let currentIdx = 0;
  for (let idx = 0; idx < sorted.length; ++idx) {
    if (sorted[idx].start.isBeforeOrEqual(results[currentIdx].end))
      results[currentIdx] = new TT(results[currentIdx].start, sorted[idx].end)
    else {
      results.push(sorted[idx]);
      ++currentIdx;
    }
  }
  return results;
}


// ^[^"\n]*?"([^"\n]*)"[^"\n]*"([^"\n]*)"


src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import { IncrementalSearch, SearchDirection, SearchOptions } from './IncrementalSearch';
import { SearchStatusBar } from './SearchStatusBar';
import * as configuration from './Configuration';
const INCREMENTAL_SEARCH_CONTEXT = 'incrementalSearch';


let status: SearchStatusBar;
let searches = new Map<vscode.TextEditor, IncrementalSearch>();

let cancellationSource = null;

let dbg = vscode.window.createOutputChannel("IncrementalSearch-extension");

// var registeredTypeCommand = false;
var context: vscode.ExtensionContext | null = null;

function registerTextEditorCommand(commandId: string, run: (editor: vscode.TextEditor, edit: vscode.TextEditorEdit, ...args: any[]) => void): void {
  context!.subscriptions.push(vscode.commands.registerTextEditorCommand(commandId, run));
}

function registerCommand(commandId: string, run: (...args: any[]) => void): void {
  context!.subscriptions.push(vscode.commands.registerCommand(commandId, run));
}

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(activationContext: vscode.ExtensionContext) {
  context = activationContext;

  vscode.commands.executeCommand('setContext', 'incrementalSearch', false);

  // if(selectionDecoration)
  //   selectionDecoration.dispose();
  // selectionDecoration = vscode.window.createTextEditorDecorationType({
  //   backgroundColor: 'rgba(0,0,255,0.5)',
  //   borderRadius: '50%',
  //   border: '1pt rgba(0,0,100,0.8) solid',
  // });
  // matchDecoration = vscode.window.createTextEditorDecorationType(configuration.get().matchStyle);
  // context.subscriptions.push(matchDecoration);

  status = new SearchStatusBar('extension.incrementalSearch.toggleCaseSensitivity', 'extension.incrementalSearch.toggleRegExp');
  context.subscriptions.push(status);
  configuration.activate();

  registerTextEditorCommand('extension.incrementalSearch.forward', (editor) => {
    advanceSearch(editor, { direction: SearchDirection.forward });
  });

  registerTextEditorCommand('extension.incrementalSearch.backward', (editor: vscode.TextEditor) => {
    advanceSearch(editor, { direction: SearchDirection.backward });
  });

  registerTextEditorCommand('extension.incrementalSearch.expand', (editor: vscode.TextEditor) => {
    advanceSearch(editor, { expand: true, direction: SearchDirection.forward });
  });

  registerTextEditorCommand('extension.incrementalSearch.backwardExpand', (editor: vscode.TextEditor) => {
    advanceSearch(editor, { expand: true, direction: SearchDirection.backward });
  });

  registerTextEditorCommand('extension.incrementalSearch.toggleRegExp', (editor: vscode.TextEditor) => {
    const search = searches.get(editor);
    if (search) {
      updateSearch(search, { useRegExp: !search.useRegExp });
    }

    context.globalState.update("useRegExp", search!.useRegExp);
  });
  registerTextEditorCommand('extension.incrementalSearch.toggleCaseSensitivity', (editor: vscode.TextEditor) => {
    const search = searches.get(editor);
    if (search) {
      updateSearch(search, { caseSensitive: !search.caseSensitive });
    }

    context.globalState.update("caseSensitive", search!.caseSensitive);
  });
  registerTextEditorCommand('extension.incrementalSearch.centerText', (editor: vscode.TextEditor) => {
    IncrementalSearch.centerText();
  });

  // registerTextEditorCommand('extension.incrementalSearch.stop', (editor) => {
  //   cancelSearch(editor);
  // });
  registerTextEditorCommand('extension.incrementalSearch.stop', (editor: vscode.TextEditor) => {
    dbg.appendLine("cancel search called");
    if (cancellationSource != null) {
      cancellationSource.cancel();
    }
    cancelSearch(editor);
  });
  vscode.window.onDidChangeActiveTextEditor(async () => {
    const search = searches.get(vscode.window.activeTextEditor!);
    if (search) {
      status.show();
      await vscode.commands.executeCommand('setContext', 'incrementalSearch', true);
      updateSearch(search, {});
    } else {
      status.hide();
      await vscode.commands.executeCommand('setContext', 'incrementalSearch', false);
    }
  });
}

function cancelSearch(editor: vscode.TextEditor) {
  let search = searches.get(editor)
  if (search)
    search.cancelSelections();
  stopSearch(editor, "stop command");
}

async function stopSearch(editor: vscode.TextEditor, reason: string, forwardCommand = '', ...args: any[]) {
  const search = searches.get(editor);
  try {
    await vscode.commands.executeCommand('setContext', 'incrementalSearch', false);
  } catch (e) { }

  if (search) {
    console.log("search stopped: " + reason);
    clearMatchDecorations(search);
    searches.delete(editor);
  }
  vscode.commands.executeCommand("cancelSelection");
  status.hide();

  try {
    if (forwardCommand)
      await vscode.commands.executeCommand(forwardCommand, args);
  } catch (e) {

  }
}

let previousSearchTerm = '';
async function doSearch(editor: vscode.TextEditor, options: SearchOptions) {
  if (searches.has(editor))
    return;

  const search = new IncrementalSearch(editor, options);
  searches.set(editor, search);
  status.update(search.searchTerm!, search.caseSensitive!, search.useRegExp!, { backward: search.direction == SearchDirection.backward });
  status.show();
  await vscode.commands.executeCommand('setContext', 'incrementalSearch', true);

  try {
    updateSearch(search, { searchTerm: 'previousText' });
    cancellationSource = new vscode.CancellationTokenSource();
    let token = cancellationSource.token;
    const searchTerm = await vscode.window.showInputBox({
      value: previousSearchTerm,
      prompt: "incremental search",
      placeHolder: "enter a search term",
      validateInput: (text: string) => {
        const result = updateSearch(search, { searchTerm: text });
        return result.error;
      }
    }, token);
    cancellationSource.dispose();
    cancellationSource = null;
    if (search.searchTerm) {
      previousSearchTerm = search.searchTerm;
    }

    if (searchTerm !== undefined && search.searchTerm) {
      stopSearch(editor, 'complete');
    } else {
      if (search)
        search.cancelSelections();
      stopSearch(editor, 'cancelled by user');
    }
  } catch (e) {
    console.error(e);
  }
}

function advanceSearch(editor: vscode.TextEditor, options: SearchOptions) {
  const search = searches.get(editor);
  if (!search) {
    const useRegExp = context!.globalState.get<boolean>("useRegExp");
    const caseSensitive = context!.globalState.get<boolean>("caseSensitive");
    if (useRegExp !== undefined)
      options.useRegExp = useRegExp;
    if (caseSensitive !== undefined)
      options.caseSensitive = caseSensitive;
    doSearch(editor, options);
  } else {
    if (search.searchTerm == '') {
      if (previousSearchTerm != '') {
        updateSearch(search, { searchTerm: previousSearchTerm });
      }
    } else {
      const results = search.advance(options);
      status.update(search.searchTerm!, search.caseSensitive!, search.useRegExp!);
      updateMatchDecorations(search, results);
    }
  }
}

/** If subgroups are matched, then display a decoration over the entire
 * matching range to help the user identify how the regexp is working
 * */
function updateMatchDecorations(search: IncrementalSearch, results: { matchedRanges: vscode.Range[], matchedGroups: boolean }) {
  if (configuration.get().selectionDecoration)
    search.getEditor().setDecorations(configuration.get().selectionDecoration, []);

  if (configuration.get().styleMatches == 'always' || (results.matchedGroups && configuration.get().styleMatches == 'multigroups'))
    search.getEditor().setDecorations(configuration.get().matchDecoration!, results.matchedRanges);
  else
    search.getEditor().setDecorations(configuration.get().matchDecoration!, []);
}

function clearMatchDecorations(search: IncrementalSearch) {
  search.getEditor().setDecorations(configuration.get().matchDecoration!, []);
}

function updateSearch(search: IncrementalSearch, options: SearchOptions): { error?: string } {
  if (!search)
    return {};

  try {
    const results = search.update(options);
    status.update(search.searchTerm!, search.caseSensitive!, search.useRegExp!);
    updateMatchDecorations(search, results);
    return {};
  } catch (e) {
    clearMatchDecorations(search);
    status.update(search.searchTerm!, search.caseSensitive!, search.useRegExp!);
    if (e instanceof SyntaxError) {
      status.indicateSyntaxError();
      return { error: e.message };
    }
    else
      console.error(e);
    return { error: 'Unknown error' }
  }
}





// this method is called when your extension is deactivated
export function deactivate() { }
